lib/OpenQA/Worker/Job.pm

```diff
@@ -0,0 +1,1245 @@
+    return $self->client->send(
+        post          => "jobs/$job_id/set_done",
+        params        => $params,
+        non_critical  => 1,
+        ignore_errors => 1,
+        callback      => sub {
+            $self->_set_status(stopped => {reason => $reason});
+        },
+    );
+
+    # note: The worker itself will react the the changed status and unassign this job object
+    #       from its current_job property and will clean the pool directory.
+}
+
+sub is_backend_running {
+    my ($self) = @_;
+
+    my $engine = $self->engine;
+    return !!0 unless $engine;
+    return !exists $engine->{child} ? !!0 : $engine->{child}->is_running;
+}
+
+sub start_livelog {
+    my ($self) = @_;
+
+    return undef unless $self->is_backend_running;
+
+    my $pooldir         = $self->worker->pool_directory;
+    my $livelog_viewers = $self->livelog_viewers + 1;
+    if ($livelog_viewers == 1) {
+        log_debug('Starting livelog');
+        open(my $fh, '>', "$pooldir/live_log") or die "Cannot create live_log file";
+        close($fh);
+    }
+    else {
+        log_debug("New livelog viewer, $livelog_viewers viewers in total now");
+    }
+    $self->{_livelog_viewers} = $livelog_viewers;
+    $self->upload_results_interval(undef);
+    $self->_upload_results();
+}
+
+sub stop_livelog {
+    my ($self) = @_;
+
+    return unless $self->is_backend_running;
+
+    my $pooldir         = $self->worker->pool_directory;
+    my $livelog_viewers = $self->livelog_viewers;
+    if ($livelog_viewers >= 1) {
+        $livelog_viewers -= 1;
+    }
+    if ($livelog_viewers == 0) {
+        log_debug('Stopping livelog');
+        unlink "$pooldir/live_log";
+    }
+    else {
+        log_debug("Livelog viewer left, $livelog_viewers remaining");
+    }
+    $self->{_livelog_viewers} = $livelog_viewers;
+    $self->upload_results_interval(undef);
+}
+
+# posts the setup status
+# note: Called by the engine while preparing the startup so the job is not considered dead by the
+#       web UI if this takes longer.
+sub post_setup_status {
+    my ($self) = @_;
+
+    # allow the worker to stop when interrupted during setup
+    return 0 if ($self->is_stopped_or_stopping);
+
+    my $client    = $self->client;
+    my $job_id    = $self->id;
+    my $worker_id = $client->worker_id;
+    if (!defined $worker_id || !defined $job_id) {
+        die 'attempt to post setup status without worker and/or job ID';
+    }
+
+    log_debug("Updating status so job $job_id is not considered dead.");
+    $client->send(
+        post     => "jobs/$job_id/status",
+        json     => {status => {setup => 1, worker_id => $worker_id}},
+        callback => 'no',
+    );
+    return 1;
+}
+
+sub _calculate_upload_results_interval {
+    my ($self) = @_;
+
+    my $interval = $self->upload_results_interval;
+    return $interval if $interval;
+
+    if ($self->livelog_viewers >= 1) {
+        $interval = 1;
+    }
+    else {
+        $interval = 10;
+    }
```

Comment:
`$interval = $self->livelog_viewers >= 1 ? 1 : 10` -> but you can do that in a follow-up PR as well